builder/builder.go

```diff
@@ -96,10 +94,13 @@ type ImageCache interface {
 type Image interface {
 	ImageID() string
 	RunConfig() *container.Config
+	MarshalJSON() ([]byte, error)
+	NewChild(child image.ChildConfig) *image.Image
```

Comment:
`AddHistory` wasn't very clear name. I meant a function that appends to both history and rootfs array of the base image config. `img.OSFeatures/OSVersion` come from the base input image that is passed in. It is also missing `Config/ContainerConfig` that could be either added by this function on set manually by the caller as they carry no extra logic other than just setting a propery on a struct.